### ISSUE TITLE
While writing formmula in an excel sheet NVDA does not report the selected cell address from current/another sheet/workbook

### ISSUE THREAD

**Issue by saidarshan786** [nvaccess/nvda#6067]
While writing a formula in excel sheet NVDA does not report the selected cell address inside the parenthesis which makes the formula writing process really challenging, user had to first note down all the cell addresses will be used inside the formula & then have to generate the formula from scratch as during the formula edit mode NVDA does not report the selected address.

For e.g. user want to write a simple formula on cell A1 in sheet2 which will calculate the length of a cell A1 from sheet1.
Required formula =LEN(SHEET1!A1)
While entering above formula after the parenthesis if we move our focus to sheet1 then NVDA does not report the selected cell/sheet hence user have to type everything on their own.

**Comment by derekriemer** [comment]
Confirmed


On 6/14/2016 8:24 AM, saidarshan786 wrote:
>
> While writing a formula in excel sheet NVDA does not report the 
> selected cell address inside the parenthesis which makes the formula 
> writing process really challenging, user had to first note down all 
> the cell addresses will be used inside the formula & then have to 
> generate the formula from scratch as during the formula edit mode NVDA 
> does not report the selected address.
>
> For e.g. user want to write a simple formula on cell A1 in sheet2 
> which will calculate the length of a cell A1 from sheet1.
> Required formula =LEN(SHEET1!A1)
> While entering above formula after the parenthesis if we move our 
> focus to sheet1 then NVDA does not report the selected cell/sheet 
> hence user have to type everything on their own.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/nvaccess/nvda/issues/6067>, or mute the thread 
> <https://github.com/notifications/unsubscribe/AFGiveG_TNB_cYkdjy2UaxsQSxQsYFzzks5qLrm0gaJpZM4I1YP6>.
>

-- 
------------------------------------------------------------------------


    Derek Riemer

  * Department of computer science, third year undergraduate student.
  * Proud user of the NVDA screen reader.
  * Open source enthusiast.
  * Member of Bridge Cu
  * Avid skiier.

Websites:
Honors portfolio <http://derekriemer.com>
Awesome little hand built weather app! 
<http://django.derekriemer.com/weather/>

email me at derek.riemer@colorado.edu <mailto:derek.riemer@colorado.edu>
Phone: (303) 906-2194

**Comment by jcsteh** [comment]
You wrote:

> which makes the formula writing process really challenging, user had to first note down all the cell addresses will be used inside the formula & then have to generate the formula from scratch

While it's certainly true that we don't read the addresses when you use the cursor keys, I'm not clear as to why this makes things "less challenging" and avoids keeping track of addresses. Whether you type "a2" or use the arrow keys to select a2, you still have to know you want the a2 cell. How does speaking the cell address as you propose solve this? I guess it does make it slightly easier for sheet names, but you still have to remember which cell addresses you want, correct?

**Comment by ruifontes** [comment]
Why not reading the sheet name, if different of actual, cell coordinates and 
contents?

Rui

-----Mensagem Original----- 
De: James Teh
Data: quinta-feira, 23 de junho de 2016 20:41
Para: nvaccess/nvda
Assunto: Re: [nvaccess/nvda] While writing formmula in an excel sheet NVDA 
does not report the selected cell address from current/another 
sheet/workbook (#6067)



You wrote:



which makes the formula writing process really challenging, user had to 
first note down all the cell addresses will be used inside the formula & 
then have to generate the formula from scratch



While it's certainly true that we don't read the addresses when you use the 
cursor keys, I'm not clear as to why this makes things "less challenging" 
and avoids keeping track of addresses. Whether you type "a2" or use the 
arrow keys to select a2, you still have to know you want the a2 cell. How 
does speaking the cell address as you propose solve this? I guess it does 
make it slightly easier for sheet names, but you still have to remember 
which cell addresses you want, correct?

-

**Comment by jcsteh** [comment]
Well, the reporter only proposed reading the address, not the contents. Contents is tricky because the formula field still has focus here. It'd be good to get some idea of the visual user experience here.

**Comment by jcsteh** [comment]
@QChristensen, can you comment on the visual UX here?

**Comment by Qchristensen** [comment]
Sorry, not sure how I missed this one before!  When typing a formula (in enter mode rather than edit mode), if you press the arrow keys to move around, Excel puts a coloured border and light coloured shading on the cell that you have arrowed to and writes either the cell name or address in the formula in the same colour.  A sighted user could then see which cell was highlighted and that it was the one they wanted.  I agree somehow automatically reading out the formula as well as the contents of the other cell would get confusing.  Perhaps this action could move the review cursor, and so NVDA+delete could then be used to read the contents of the cell the user has arrowed to?  The focus never actually leaves the cell being edited though.

Just for completeness (and getting off track a little), Excel also highlights any other cells referenced in the formula, each in a different colour.

Also, using Excel 2016 and NVDA 2017.1, if I press left or right arrows to move within the same row while editing a formula, NVDA reads "blank".  If I press up or down, NVDA reads the formula as it changes, effectively telling me the current coordinates.  Nothing at all is read if I hold down shift to select multiple cells (the selection occurs and is added to the formula).

**Comment by Adriani90** [comment]
I agree this would help alot, also optionally reading also the contents of the cell while arrowing through the cell editing field would really accelerate things. At least for me, when I am editing complex financial models, I must remember many cells at once. Having the content spoken when using arrow keys to edit a formula would help alot because I would know how far I must navigate upwards or downwards to find my desired cell for the formula. Only the coorinates do not help very much. I think this is quite tricky to implement but it would really change things positively.

**Comment by Qchristensen** [comment]
Tested today using:
NVDA 2023.1.0.27913   (No add-ons)
Windows 11 (64-bit) Version: 22H2, Build: 22621.1702 
Office 365 (64-bit) Version: 16.0.16327.20248 

This is still present, although with some improvement:

If you are typing a formula (eg, press = on a cell), as you navigate around the worksheet, NVDA will read the cell coordinates.  If the cell is not blank, it will also read the formula value (the results of the formula in that cell, not the formula itself).

With NVDA set to use UIA in Excel:
- NVDA DOES read the current cell in other worksheets.  If the cell has contents, it will read the contents rather than the coordinates
- NVDA does not read the current cell in other workbooks:  Press ALT+TAB to move to another open Workbook.  The formula bar indicatses ='[Workbook]Worksheet'!:$A$1 - however NVDA only repeats the current cell in the original workbook.

With NVDA set to not use UIA in Excel (the default):
- While editing a formula, if I try to change worksheets, NVDA shows the new worksheet visually for a moment, then jumps back to the original worksheet.  As I navigate around the current worksheet after this, NVDA only reports the formula value, not cell reference ("formula value 0" for blank cells).
- When trying to select a cell in another workbook, the formula bar indicates the current cell which will be added to the formula, however NVDA is completely silent.  If I alt tab back to the editing workbook, the cell from the other worksheet is selected in the formula, ready to insert (as also happens if I press enter from the other worksheet).

**Comment by Qchristensen** [comment]
No update, although this is still an issue users report experiencing.
